Introducing a new barriers API

carlocaione opened the pull request:
Zephyr is missing a proper API for data memory barriers.

We currently have some kind of ARM-centric arch-specific macros scattered around called (mostly) `__DMB()` or `__DSB()`.

In general the situation is confusing because:
- There is no a standard nomenclature for the macros. Sometimes they are called `__DMB()`, sometimes `dmb()`, sometimes `cpu_dmb()`, sometimes something arch-specific entirely, etc...
- Each architecture is doing something different. On ARM those are falling back to CMSIS, on ARM64 those are implemented in assembly, on several architecture they do not have an implementation at all, etc... See for example:
 https://github.com/zephyrproject-rtos/zephyr/blob/9ad78eb60cfa629c1caf7217dbbf7adc5d747011/subsys/testsuite/include/zephyr/test_asm_inline_gcc.h#L26-L44
- Since each architecture is doing something different we are starting to see something like this https://github.com/zephyrproject-rtos/zephyr/pull/57363#discussion_r1180049262 and https://github.com/zephyrproject-rtos/zephyr/pull/57363/commits/8d3718e4ac2795be13ee3b5b16ddd851d009369c where the drivers compiled for different architectures needs to carry boilerplate code to deal with this confusion
- Core code is missing this in SMP code, for example: https://github.com/zephyrproject-rtos/zephyr/blob/ca6c08f960d83265497af261c5eda2e782c47b26/kernel/include/kswap.h#L38-L41

This patch is meant to cleanup a bit the situation, introducing a proper API like it was done for the atomic API already in place. When possible we try to leverage the compiler and the built-in functions for memory model aware atomic operations like it was suggested already in #42625

We use arch-specific implementation only on ARM and ARM64 and use the built-in implementation on all the others.

carlocaione committed:
barriers: Introduce barrier operations

Introduce a new API for barrier operations starting with a general
skeleton and the implementation for barrier_data_memory_fence_full().

Select a built-in or an arch-based implementation according to new
Kconfig symbols CONFIG_BARRIER_OPERATIONS_BUILTIN and
CONFIG_BARRIER_OPERATIONS_ARCH.

The built-in implementation falls back on the compiler built-in
function using __ATOMIC_SEQ_CST as it is done for the atomic APIs
already.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: arm: Use CMSIS instead of built-in functions

For ARM (AArch32) we fall back on the CMSIS implementation that it is
supposedly more refined.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: Move __DMB() to the new API

Remove the arch-specific ARM-centric __DMB() macro and use the new
barrier API instead.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: Introduce barrier_dsync_fence_full()

Some architectures have a kind of fence that is "stronger" than the
regular data memory barrier. Introduce a new API for that.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: arm: Use CMSIS for DSB

As done for DMB, use CMSIS for DSB as well.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: arm64: Move to arch-specific barriers

The compiler is not able to emit a proper DSB operation for ARM64. Move
to the arch-specific implementation and use assembly code instead.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: Move __DSB() to the new API

Remove the arch-specific ARM-centric __DSB() macro and use the new
barrier API instead.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: Introduce barrier_isync_fence_full()

Some architectures have a special barrier instruction that is used to
ensure that any previously executed context-changing operations, such as
writes to system control registers, have completed by the time the
instruction is done.

In the ARM world this is called ISB. For all the other architectures
falls back to a compiler barrier.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: arm: Use CMSIS for ISB

As done before, falls back onto CMSIS also for ISB.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: arm64: Implement arch-specific barrier op

For barrier_isync_fence_full() use the assembly code to generate a
proper ISB instruction.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione committed:
barriers: Move __ISB() to the new API

Remove the arch-specific ARM-centric __ISB() macro and use the new
barrier API instead.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

microbuilder commented on arch/arm/core/aarch32/cortex_m/fault.c:
```diff
@@ -710,14 +711,14 @@ static inline bool z_arm_is_synchronous_svc(z_arch_esf_t *esf)
 	SCB->CCR |= SCB_CCR_BFHFNMIGN_Msk;
-	__DSB();
-	__ISB();
+	barrier_data_synchronization_fence_full();
+	barrier_instruction_synchronization_fence_full();
```
Would ...

```suggestion
	barrier_dsync_fence_full();
	barrier_isync_fence_full();
```

and `barrier_dmem_fence_full`(etc.) be more concise but still clear?

carlocaione replied:
yup, sounds fine to me. I'll change that.

gmarull replied:
+1 to shorter names

carlocaione replied:
done

lorc commented on drivers/xen/events.c:
```diff
@@ -219,7 +220,7 @@ static void events_isr(void *data)
 	vcpu->evtchn_upcall_pending = 0;
 
-	dmb();
+	barrier_data_memory_fence_full();
```
I believe, we need `barrier_data_synchronization_fence_full` in this case, as this part synchronizes access to vCPU structures between Zephyr and Hypervisor. I am not sure that GCC's idea of synchronization is valid for inter-VM communications, so I believe we need proper `dsb sy` there.

carlocaione replied:
Previously `dmb()` was expanding to `dmb sy`. Now with this PR, GCC is expanding `barrier_data_memory_fence_full()` to `dmb ish`.

Are you suggesting that using `dmb ish` instead `dmb sy` is the problem here? Or that we should use `dsb` instead of `dmb`? Because in the latter case the problem was present even before this PR so that should be fixed in a different PR possibly (or at least commit).

lorc commented on drivers/xen/gnttab.c:
```diff
@@ -86,7 +87,7 @@ static void gnttab_grant_permit_access(grant_ref_t gref, domid_t domid,
 	gnttab.table[gref].domid = domid;
 	/* Need to be sure that gfn and domid will be set before flags */
-	__DMB();
+	barrier_data_memory_fence_full();
```
I believe, we need `barrier_data_synchronization_fence_full` in this case, as this part synchronizes access to grant references tables between Zephyr and Hypervisor. I am not sure that GCC's idea of synchronization is valid for inter-VM communications, so I believe we need proper `dsb sy` there.

carlocaione replied:
idem as in my previous comment.

lorc reviewed (commented):
Okay, so my comments regarding `dmb` in xen-related code are not exactly vaid, but now I have a more broad question: are there any guarantees that `__atomic_thread_fence(__ATOMIC_SEQ_CST);` on ARM64 will be expanded into `dmb ish`? I do understand that GCC right now does exactly this. But all that GCC documentation guarantee is that:

1. It will "approximately match the requirements for the C++11 memory model" (https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html)
2. "This built-in function acts as a synchronization fence between **threads** based on the specified memory order." 

This is fine in the cases when Zephyr uses GCC built-ins to synchronize memory accesses between threads. But I am not sure if this is okay for cases when are talking to hardware or doing inter-VM communication. If GCC does not guarantee to use `dmb ish` exactly, those cases might break. Theoretically, GCC might use some other synchronization technique that provides "approximate requirements for the C++11 memory model" for synchronization between **threads**. 

Also, `dmb ish` is weaker than `dmb sy`.  Again, this is fine for synchronization between threads because "This architecture assumes that all PEs that use the same operating system or hypervisor are in the same Inner Shareable shareability domain." (ARM DDI 0487.a page B2-197). But this might create issues for code that is working with hardware, because this code might rely on full system  synchronization. 

As a bottomline, I am proposing to consider implementing  `barrier_data_memory_fence_full()` as `	__asm__ volatile ("dmb sy" ::: "memory");` on ARM.

andyross commented:
[jumping in, driving by, running away...]

> are there any guarantees that __atomic_thread_fence(__ATOMIC_SEQ_CST); on ARM64 will be expanded into `dmb ish`?

Pedantically, no.  The toolchain/platform is free to do whatever it wants to guarantee the requirements of the spec.  In an obvious and specious case: if it knows it's running on an in-order single core machine, it can elide the instruction.  In practice yes, because the toolchain is built for modern systems with parallel CPUs in mind.

And the language in the standard doesn't talk about separate, non-CPU bus clients because those things aren't part of the C++ language specification.  It just doesn't treat them.

And yeah, that's annoying, which is sort of the reason the OS/platform layer needs to step in and make promises the toolchain isn't willing to.  So the point to Zephyr doing this is that **Zephyr** is making this promise, and any bugs that result in the generated code end up being Zephyr bugs, even if all we're doing by default is just forwarding to the toolchain.

[And having said that, let me flee before I get drawn into a standards discussion]

gmarull commented on include/zephyr/sys/barrier.h:
```diff
@@ -0,0 +1,24 @@
+
+#ifdef __cplusplus
+} /* extern "C" */
+#endif
```
Doxygen docstrings should be present in this header, which is the only one included by API clients. Others are just implementation details.

carlocaione replied:
what about something like:
```c
#if defined(CONFIG_BARRIER_OPERATIONS_ARCH) && !defined(__DOXYGEN__)
/* Empty */
#else
#include <zephyr/sys/barrier_builtin.h>
#endif
```

gmarull replied:
That doesn't solve the problem, I was talking about making top header the source of docstrings. Maybe the following pattern can be followed instead:

`zephyr/sys/barrier.h`:

```c
/**
 * @brief Full/sequentially-consistent data memory barrier.
 *
 * This routine acts as a synchronization fence between threads and prevents
 * re-ordering of data accesses instructions across the barrier instruction.
 */
static ALWAYS_INLINE void barrier_dmem_fence_full(void)
{
    z_barrier_dmem_fence_full();
}
```

`zephyr/sys/barrier_builtin.h` (or arch specific headers):

```c
static ALWAYS_INLINE void z_barrier_dmem_fence_full(void)
{
	__atomic_thread_fence(__ATOMIC_SEQ_CST);
}
```

We could also introduce something like this to ease unit testing (ie, an additional choice to builtin/arch) https://www.keil.com/pack/doc/CMSIS/Core/html/group__NVIC__gr.html#gadc48b4ed09386aab48fa6b9c96d9034c

carlocaione replied:
done

gmarull commented on include/zephyr/sys/barrier_builtin.h:
```diff
@@ -0,0 +1,41 @@
+extern "C" {
+#endif
+
+/* Included from <barrier.h> */
```
we should probably add an include guard like:

```c
#ifndef ZEPHYR_INCLUDE_SYS_BARRIER_H_
#error Please include <zephyr/sys/barrier.h>
#endif
```

gmarull commented on include/zephyr/sys/barrier_builtin.h:
```diff
@@ -0,0 +1,41 @@
+ * This routine acts as a synchronization fence between threads and prevents
+ * re-ordering of data accesses instructions across the barrier instruction.
+ */
+static ALWAYS_INLINE void barrier_dmem_fence_full(void)
```
`ALWAYS_INLINE` -> `#include <zephyr/toolchain.h>`

gmarull commented on kernel/Kconfig:
```diff
@@ -452,6 +452,21 @@ config SYSTEM_WORKQUEUE_NO_YIELD
 
 endmenu
 
+menu "Barrier Operations"
+config BARRIER_OPERATIONS_BUILTIN
+	bool
+	help
+	  Use the compiler builtin functions for barrier operations. This is
+	  the preferred method. However, support for all arches in GCC is
+	  incomplete.
+
+config BARRIER_OPERATIONS_ARCH
+	bool
+	help
+	  Use when there isn't support for compiler built-ins, but you have
+	  written optimized assembly code under arch/ which implements these.
+endmenu
```
`choice`?

carlocaione replied:
yeah, choice is really annoying because you cannot do `select WHATEVER`  otherwise the Kconfig will complain with `warning: the choice symbol WHATEVER  is selected by the following symbols, but select/imply has no effect on choice symbols`

gmarull commented on include/zephyr/arch/arm/aarch32/barrier_arm.h:
```diff
@@ -0,0 +1,22 @@
+#include <zephyr/arch/arm/aarch32/cortex_a_r/cmsis.h>
+#endif
+
+/* Included from <sys/barrier.h> */
```
ditto: prevent direct includes

carlocaione replied:
uhm, how CMSIS should be included?

gmarull commented on include/zephyr/arch/arm64/barrier_arm64.h:
```diff
@@ -0,0 +1,21 @@
+#ifndef ZEPHYR_INCLUDE_BARRIER_ARM64_H_
+#define ZEPHYR_INCLUDE_BARRIER_ARM64_H_
+
+/* Included from <sys/barrier.h> */
```
ditto

gmarull commented on include/zephyr/arch/arm64/barrier_arm64.h:
```diff
@@ -0,0 +1,21 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_BARRIER_ARM64_H_
```
header name adds redundancy as well

```
include/zephyr/arch/arm64/barrier_arm64.h
                    ^             ^
```

de-nordic reviewed (dismissed):
Changes to flash drivers look ok

lorc reviewed (dismissed):
I am fine with changes to the xen code and with implementation of barrier functions on arm.

gmarull reviewed (dismissed):
LGTM, thanks!

carlescufi commented:
Added DNM since, although this was already presented in the Architecture WG, we should let more eyes on it before we merge it.

stephanosio commented on include/zephyr/sys/barrier_builtin.h:
```diff
@@ -0,0 +1,31 @@
+
+static ALWAYS_INLINE void z_barrier_dmem_fence_full(void)
+{
+	__atomic_thread_fence(__ATOMIC_SEQ_CST);
```
This is a GCC built-in and will only work with GCC (and, at most, with Clang which intends to be compatible with GCC).

Since Zephyr kernel needs to be able to build with toolchains other than GCC, there should be an additional backend using the C11 `atomic_thread_fence`.

carlocaione replied:
but shouldn't this be a problem also for the atomic built-ins in https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/atomic_builtin.h ?

carlocaione replied:
what about using
```c
#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L
```
to switch to `atomic_thread_fence` in the same backend?

carlocaione replied:
Added https://github.com/zephyrproject-rtos/zephyr/pull/57439/commits/7da73c20668eb3ac77f0a36ec0178cfab675bc2e . Is this ok?

stephanosio replied:
Added review comment here: https://github.com/zephyrproject-rtos/zephyr/pull/57439#discussion_r1187394647

stephanosio reviewed (commented):
(Not requesting changes, just a comment)

IMHO, the memory barrier API being added in this PR seems to be too basic and limited to be of practical use -- for instance, it does not separate read (load) and write (store) barriers, nor does it make a clear distinction between the non-SMP and SMP (cache) barriers.

Should we define a more complete API at this time so that we do not need to make breaking API changes in the future to support the aforementioned features?

carlocaione commented:
@stephanosio 

> for instance, it does not separate read (load) and write (store) barriers

Do we have an actual usage of that in tree right now? Anyway, no objection to add that as well.

> nor does it make a clear distinction between the non-SMP and SMP (cache) barriers.

This was discussed in the very first implementation of this and it was kinda ruled out for the moment, see https://github.com/zephyrproject-rtos/zephyr/pull/41977#discussion_r788058472 . Again, no strong objection to add that back.

andyross commented:
FWIW, I'd personally prefer we start simple here.  Lockless algorithms are hard even with full barriers, they only get worse with separate read/write barriers like ARM has, and it gets even worse still when you have to reason using the produce/consume metaphor favored by the C++ standard (and thus toolchain API).

If the OS is going to have to make the promise, let's have it make the simple promise.  Adding features are optimizations we can do later.

And as far as example in-tree, there's a chicken/egg problem.  Right now we have a handful of spots where we should have barriers but don't (the core switch code in the scheduler, winstream are ones I know about), but are being saved by the fact that these areas are next to other synchronizing instructions, or are running on architectures that are in-order.

And as far as cache-awareness, that's sort of always going to be platform dependent.  The cache incoherence seen with xtensa, for example, really can't be treated with a barrier API at all and has to involve app code knowing where to flush (and also how to arrange the items in memory as there's no bytewise cache state, it all goes as full lines).

andyross reviewed (dismissed):
Sorry, forgot vote.  This looks good to me as-is.  For sure this is likely to evolve as new users arrive, but we should at least have the simple API in tree for now.

de-nordic reviewed (dismissed):
Flash drivers look ok.

carlocaione committed:
barriers: builtin: Check for GCC-ism

__atomic_thread_fence() is GCC-ism. When supported by the compiler
(C11), use atomic_thread_fence() instead.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

stephanosio commented on include/zephyr/sys/barrier_builtin.h:
```diff
@@ -0,0 +1,55 @@
+
+static ALWAYS_INLINE void z_barrier_dmem_fence_full(void)
+{
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ < 201112L
+	/* C99*/
+	__atomic_thread_fence(__ATOMIC_SEQ_CST);
```
This really has nothing to do with C99. Since `__atomic_thread_fence` is a GCC built-in/GNU extension, it should check for `__GNUC__` here, and otherwise assume the C11 stdatomic.

```suggestion
#if defined(__GNUC__)
	/* C99*/
	__atomic_thread_fence(__ATOMIC_SEQ_CST);
```

stephanosio reviewed (dismissed):
Compiler agnosticity issue was addressed.

evgeniy-paltsev reviewed (changes requested):
Hi,

I'm a bit concerned that we mix all potential kinds of HW barriers here.

And especially I'm concern about mixing non-SMP and SMP data barriers.

Various architecture may need to implement them in a diffenet way, i.e:

- in some cases we may want to have it both
- in some cases we may want to have it both, but implemented with different instructions
- in case of single core Zephyr we most likely want to have non-SMP barriers only (and implement SMP barriers as a stab)
- in case of some architectures we may only need SMP barriers

Proper separarion of the non-SMP and SMP barriers allow to handle these cases on the architecture level. Current API only allows to handle it with `#ifdef`'s around the places where we use new barrier API.

And we already have multiple of such places, i.e.:

```
#if defined __CORTEX_M && (__CORTEX_M == 4U)
	barrier_dsync_fence_full();
#endif
```

Currently we use this API mostly in architecture code and in architecture / SoC - specific drivers. However, we expose this API to all drivers. And with adding support of more and more DMA-enabled peripherals (where non-SMP barriers might be required) this aproach will lead us to `#ifdef` chaos in generic drivers.

carlocaione commented:
@evgeniy-paltsev hello,

This PR is trying to tidy up a situation that we already have in the Zephyr code without adding too much (for now) unnecessary complexity.

You can see this as an API formalization where we have:
- `DSB()` -> `barrier_dsync_fence_full()`
- `DMB()` -> `barrier_dmemory_fence_full()`
- `ISB()` -> `barrier_isync_fence_full()`

This is it basically, we are not even trying to be smart about it by deciding whether the `DMB` was used as SMP barrier or a full barrier.

If someone mixed SMP and non-SMP barriers it was done already and it is already in the tree code, we can only hope that `DSB` and `DMB` were used by the developer knowing what they were doing.

If new architecture, new code or new developers really need (and really know what they are doing) feel the need to differentiate `local barriers` vs `SMP barriers` vs `DMA barriers`, this is something that can be addresses later (if ever needed) but for now I feel like this is a first step toward a normalization of a messy situation we already have.

Please also consider that the topic is complex enough without letting the user to decide on a case-by-case basis whether an SMP vs DMA vs local barrier is needed or not and which one.

Do you have any concrete proposal about the changes you would like to see to this PR considering the current state of the sprinkled `DMB` and `DSB` instructions in the tree?

evgeniy-paltsev reviewed (dismissed):
Hi @carlocaione,

> This PR is trying to tidy up a situation that we already have in the Zephyr code without adding too much (for now) unnecessary complexity.

> You can see this as an API formalization

Indeed!

Currently we just use architecture-specific functions in architecture-specific code (whenewer if it is in `arch/`, `soc/` or in architecture-specific driver).

That has one advantage - no one will use it for another architecture (as it simply won't compile) - and if someone write architecture-specific code we can at least hope that they a bit familiar with this architecture.

As opposed to that, this PR introduce the API, and this API is available for all architectures - and it can be used in generic code like some peripheral drivers which are available for various architectures.

This API is extreamly low-level and due to that (and due to the fact that it's just renamed ARM instructions intrinsics) it's still ARM-centric - but now somehow availabe for all architectures.

And by "somehow availabe" I mean that I'm not sure if it is implemented the right way for all architectures.

I.e: at least for ARC architecture the desctiption in the API for barrier_dsync_fence_full differs from what the provided implementation is actually does.

carlocaione commented:
> Currently we just use architecture-specific functions in architecture-specific code (whenewer if it is in `arch/`, `soc/` or in architecture-specific driver).

The problem is that we are not doing that at all right now.

We are using arch-specific functions in tests creating an `ifdef` madness:

https://github.com/zephyrproject-rtos/zephyr/blob/9ad78eb60cfa629c1caf7217dbbf7adc5d747011/subsys/testsuite/include/zephyr/test_asm_inline_gcc.h#L26-L49

or in NOT-arch specific drivers:
https://github.com/zephyrproject-rtos/zephyr/blob/8d3718e4ac2795be13ee3b5b16ddd851d009369c/drivers/serial/uart_pl011.c#L211-L213

and we even need that in the general non-arch specific kernel code:
https://github.com/zephyrproject-rtos/zephyr/blob/ca6c08f960d83265497af261c5eda2e782c47b26/kernel/include/kswap.h#L38-L41

> That has one advantage - no one will use it for another architecture (as it simply won't compile) - and if someone write architecture-specific code we can at least hope that they a bit familiar with this architecture.

But the user / dev must be able to use these functions in their code without being aware of the gory architecture details or using an `ifdef` madness to filter-out non-supported architectures. I mean, the whole point of introducing a new API is indeed abstracting the hardware away. The problem is that you are assuming that these functions can only be used for arch-specific code but this is not always the case. I have downstream drivers that can be used on ARM and RISCV at the same time (and I'm not even talking about the kernel code that is used by all the architectures).

> This API is extreamly low-level and due to that (and due to the fact that it's just renamed ARM instructions intrinsics) it's still ARM-centric - but now somehow availabe for all architectures.

Do you have any better suggestion about naming? I'm really open to any constructive technical suggestion here.

> And by "somehow availabe" I mean that I'm not sure if it is implemented the right way for all architectures.

This is why, in the default fall-back, we are relying on the compiler to make the work for us of generating the (possibly most) correct implementation. For all the other architectures that require some tweaking (like ARM64) each arch can introduce its own implementation. Please also note that we are falling back onto the most restrictive option (`__ATOMIC_SEQ_CST`) for the fencing.

> I.e: at least for ARC architecture the desctiption in the API for barrier_dsync_fence_full differs from what the provided implementation is actually does.

Thanks for checking, this is really helpful. At this point I guess we have two ways of fixing this:
- Renaming the API to something more descriptive / correct of the actual implementation and far from the ARM-centric nomenclature
- Introduce a proper ARC backend

carlescufi commented:
Arch WG:

- @evgeniy-paltsev states that the API is too Arm-specific, just a remap of the Arm barrier API
- @carlocaione corrects him, the `__DMB`, `__ISB` and family are being used for multiple architectures, including RISC-V
- @carlocaione states that the only implementations are for Arm and Aarch64, and that the rest fall back on the compiler
- @evgeniy-paltsev is concerned that the API is too low-level, and he would like to see a separation between SMP barriers and IO/cache barriers

cfriedt reviewed (commented):
@carlocaione - LGTM. It looks like the generic interface could be added to `common/sys_io.h` to be enabled for RISC-V (and maybe others), unless I've missed something. And if it's an opt-in API like what was mentioned today, then it should be a no-op if architectures have not opted in, so should be a non-issue.

carlocaione commented:
@carlescufi @nashif 

This new revision is not actually introducing any new Kconfig symbol, but we are simply falling to NOP when neither `CONFIG_BARRIER_OPERATIONS_BUILTIN` nor `BARRIER_OPERATIONS_ARCH` is selected, I felt like this is more clear and we are saving one symbol.

@maintaners please, let me know for your architecture if you want to:
1. (default choice) do nothing about barriers (in that case all the barriers operations will be NOP)
2. use `BARRIER_OPERATIONS_ARCH`, that is you want to implement an arch-dependent backend for barrier operations
3. use `CONFIG_BARRIER_OPERATIONS_BUILTIN`, that is you want to fall back on the standard compiler-provided implementation given by `__atomic_thread_fence(__ATOMIC_SEQ_CST);`

If nothing is heard from you, we are going to assume your preference is for the default (safer) choice.

- [x] ARM64 (@carlocaione) - use `BARRIER_OPERATIONS_ARCH`
- [x] ARM (@microbuilder) - use `BARRIER_OPERATIONS_ARCH`
- [ ] ARC (@evgeniy-paltsev)
- [ ] MIPS (??)
- [ ] NIOS-2 (@nashif)
- [ ] POSIX (@aescolar)
- [ ] RISC-V (@fkokosinski @kgugala @tgorochowik)
- [ ] SPARC (@julius-barendt)
- [ ] X86 (@jhedberg)
- [ ] XTENSA (@nashif)

andyross commented:
FWIW: all existing Xtensa and Nios-2 hardware, and I believe ARC but will defer to experts, are in-order devices without explicit requirements for memory ordering control (though xtensa obviously has an incoherent cache which makes a huge mess of the same kinds of problems).  They should all be fine with the noop.

That said: surely the compiler builtins will emit a noop on their own anyway, no?  Are there toolchains for which the barrier operations don't work or emit incorrect code when they should do nothing?

carlocaione commented:
> That said: surely the compiler builtins will emit a noop on their own anyway, no? Are there toolchains for which the barrier operations don't work or emit incorrect code when they should do nothing?

@andyross this was decided during the arch meeting to be on the safe side and to not change behaviour for the arch that didn't have barriers before. This was also decided because @evgeniy-paltsev reported that what was generated by the compiler was not 100% aligned with the description of the API.

xingrz reviewed (approved):
LGTM. This matches my needs perfectly for compiling PL011 for RISC-V.

carlocaione commented:
> @carlocaione You can set CONFIG_BARRIER_OPERATIONS_BUILTIN for the posix arch

thank you, noted. Instead of pushing every single time (and dismissing the ACKs at the same time), I'll gather all the arch maintainers preferences and I'll push a new PR after this has been merged.

andyross reviewed (approved):
Looks like we're having a checkmark party.  Guess I should throw one up too.  Ship it.
